kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs

```diff
@@ -563,6 +564,45 @@ impl<'tcx> GotocCtx<'tcx> {
         }
     }
 
+    /// Codegen the reference to a given place.
+    /// We currently have a somewhat weird way of handling ZST.
+    /// - For `*(&T)` where `T: Unsized`, the projection's `goto_expr` is a thin pointer, so we
+    ///   build the fat pointer from there.
+    /// - For `*(Wrapper<T>)` where `T: Unsized`, the projection's `goto_expr` returns an object,
+    ///   and we need to take it's address and build the fat pointer.
+    pub fn codegen_place_ref(&mut self, place: &Place<'tcx>) -> Expr {
+        let place_ty = self.place_ty(place);
+        let projection = unwrap_or_return_codegen_unimplemented!(self, self.codegen_place(place));
+        if self.use_thin_pointer(place_ty) {
+            // Just return the address of the place dereferenced.
+            projection.goto_expr.address_of()
+        } else if place_ty == pointee_type(self.local_ty(place.local)).unwrap() {
+            // Just return the fat pointer if this is a simple &(*local).
+            projection.fat_ptr_goto_expr.unwrap()
+        } else {
+            // Build a new fat pointer to the place dereferenced with the metadata from the
+            // original fat pointer.
+            let proj_expr = projection.goto_expr;
+            let data = if proj_expr.typ().is_pointer() {
+                proj_expr
+            } else if proj_expr.typ().is_array_like() {
+                proj_expr.array_to_ptr()
+            } else {
+                proj_expr.address_of()
+            };
```

Comment:
Does this happen often enough it could use a utility fn?

Reply:
I haven't bumped into it in other places yet but I can still move this to a utility function.